### PR TITLE
Support Local Development

### PR DESCRIPTION
**Pull request by fabeschan**
hi

### CODE REVIEW

File: codalab/lib/bundle_cli.py [hunk 1]
```diff
@@ -1192,6 +1208,119 @@ def do_run_command(self, args):
...
+            Commands.Argument('-w', '--worksheet-spec', help='Operate on this worksheet (%s).' % WORKSHEET_SPEC_FORMAT, completer=WorksheetsCompleter),
+        ) + Commands.metadata_arguments([RunBundle]) + EDIT_ARGUMENTS + WAIT_ARGUMENTS,
+    )
+    def do_run_local_command(self, args):
+        client, worksheet_uuid = self.parse_client_worksheet_uuid(args.worksheet_spec)
+        args.target_spec, args.command = cli_util.desugar_command(args.target_spec, args.command)
+        targets = self.parse_key_targets(client, worksheet_uuid, args.target_spec)
+        metadata = self.get_missing_metadata(RunBundle, args)
+
+        docker_image = metadata.get('request_docker_image', 'ubuntu:14.04')
```
Comment (percyliang) [inline codalab/lib/bundle_cli.py]: I think If people are doing run-local, the `--request-docker-image` option should be required (they should be explicit about what they're doing.)  99% of the invocations will not be with `ubuntu:14.04`.

File: codalab/lib/bundle_cli.py [hunk 2]
```diff
@@ -1192,6 +1208,119 @@ def do_run_command(self, args):
...
+        if not docker_image:
+            docker_image = 'ubuntu:14.04'
+
+        uuid = generate_uuid()
+        bundle_path = '/opt/{}'.format(uuid)
```
Comment (percyliang) [inline codalab/lib/bundle_cli.py]: Should not be `/opt`, but should be a flag that's passed in by the user.  Maybe default to the current directory?

File: codalab/lib/bundle_cli.py [hunk 3]
```diff
@@ -1192,6 +1208,119 @@ def do_run_command(self, args):
...
+        container_id = dc.start_container(bundle_path, uuid, command, docker_image, request_network, dependencies)
+        print >>self.stdout, '===='
+        print >>self.stdout, 'ContainerID: ', container_id
+        print >>self.stdout, 'Local Bundle ID: ', uuid
+        print >>self.stdout, 'You can find local bundle contents in: ', bundle_path
+        print >>self.stdout, '===='
+
+    @Commands.command(
+        'run-image',
+        help='Create a bundle by running a program bundle on an input bundle.',
```
Comment (percyliang) [inline codalab/lib/bundle_cli.py]: Needs to be updated to be more specific.

File: codalab/lib/bundle_cli.py [hunk 4]
```diff
@@ -1192,6 +1208,119 @@ def do_run_command(self, args):
         self.wait(client, args, new_bundle['uuid'])
 
     @Commands.command(
+        'run-local',
+        help='Run a run bundle locally.',
```
Comment (percyliang) [inline codalab/lib/bundle_cli.py]: Explain much more.

File: codalab/lib/bundle_cli.py [hunk 5]
```diff
@@ -1192,6 +1208,119 @@ def do_run_command(self, args):
...
+                    ' '.join(['bash', '-c', '; '.join(docker_commands)])
+            )
+        os.system(cli_command)
+
+    @Commands.command(
+        'commit-image',
+        help='Create a bundle by running a program bundle on an input bundle.',
+        arguments=(
+            Commands.Argument('--container', help='Container to commit.', required=True),
+            Commands.Argument('--repository', help='Repository to commit to. E.g: codalabtest-on.azurecr.io/ubuntu', required=True),
```
Comment (percyliang) [inline codalab/lib/bundle_cli.py]: Use dockerhub as an example

File: codalab/lib/bundle_cli.py [hunk 6]
```diff
@@ -1192,6 +1208,119 @@ def do_run_command(self, args):
...
+                    ' '.join([ '-v {}'.format(v) for v in volume_bindings ]),
+                    ' '.join([ '--device {}'.format(d) for d in devices ]),
+                    docker_image,
+                    ' '.join(['bash', '-c', '; '.join(docker_commands)])
+            )
+        os.system(cli_command)
+
+    @Commands.command(
+        'commit-image',
+        help='Create a bundle by running a program bundle on an input bundle.',
```
Comment (percyliang) [inline codalab/lib/bundle_cli.py]: More explanation.

File: codalab/lib/bundle_cli.py [hunk 7]
```diff
@@ -1192,6 +1208,119 @@ def do_run_command(self, args):
...
+            Commands.Argument('--repository', help='Repository to commit to. E.g: codalabtest-on.azurecr.io/ubuntu', required=True),
+        )
+    )
+    def do_commit_image_command(self, args):
+        cli_command = 'docker commit {} {}'.format(args.container, args.repository)
+        os.system(cli_command)
+
+    @Commands.command(
+        'push-image',
+        help='Create a bundle by running a program bundle on an input bundle.',
```
Comment (percyliang) [inline codalab/lib/bundle_cli.py]: Update

File: codalab/lib/bundle_cli.py [hunk 8]
```diff
@@ -1192,6 +1208,119 @@ def do_run_command(self, args):
...
+        request_network = None
+        dependencies = [
+            (u'{}'.format(key), u'/{}_dependencies/{}'.format(uuid, key)) for key, target in self.parse_target_specs(args.target_spec)
+        ]
+
+        # Set up a directory to store the bundle.
+        remove_path(bundle_path)
+        os.mkdir(bundle_path)
+
+        for path1, path2 in dependencies:
```
Comment (percyliang) [inline codalab/lib/bundle_cli.py]: use `name` and `child_path` (I think) to maintain consistency with code elsewhere.

**Review by percyliang — CHANGES_REQUESTED** [review]
This functionality is great!

- In general, more documentation would be nice.
- Can you update test-cli.py to test this functionality (which is also nice since it tells us what commands to use).
- When we add new CLI commands, we should really think through whether this is the best possible design and usage.  We should have several people test it and give comments before committing to it.
- In particular, I'm worried that we're adding too many functions at the top-level.  In some sense, `run-local` and `run-image` behave a lot like `run`, so those could be flags inside `run`.  Of course that has the danger of making `run` too bloated.  But this is something we should think about more and discuss.
- Usability: Before and after each of the commands runs, there should be more stuff printed to stdout saying what's going on and what to do next.
- Also, please use more meaningful commit messages.

**Comment by fabeschan** [comment]
Hi @percyliang, thanks for looking through the changes. These are all good points, and the requested changes all make sense. However like you suggested let's hold a deeper discussion, think things through and settle into a more refined design before committing work to this.

Perhaps I can put together a google doc and we can take down some notes during upcoming meetings?

**Comment by percyliang** [comment]
Sure.  I think we should aim to wrap this up this week (don't think we're
that far off).  And while we should think about it, I think there's no
substitute for rolling it out and getting people to try it (as long as it's
understood that this is a beta feature, we can still change the interface
after it goes into develop).

**Comment by fabeschan** [comment]
https://docs.google.com/document/d/1qELOtPOU6vbA2FtgOMmuBmRrVDGnNRszIgMW8s3XZl4/edit#heading=h.uuvhzjo8mk0y

File: codalab/lib/bundle_cli.py (hunk 2 above)
Comment (fabeschan) [reply codalab/lib/bundle_cli.py]: Would it be better for bundle paths to be set in config? Having local bundles in one directory could make periodic clean-ups easier. If so, what path would be a good default?

File: codalab/lib/bundle_cli.py (hunk 1 above)
Comment (fabeschan) [reply codalab/lib/bundle_cli.py]: OK. I will require request-docker-image when running in local mode

**Comment by fabeschan** [comment]
> In some sense, run-local and run-image behave a lot like run, so those could be flags inside run.

It makes sense to absorb run-local into run with a flag, but I'm feeling a bit queasy doing the same for run-image, which is meant for editing images (with some bells added to facilitate testing) but not for producing run bundles (with their contents).

I would prefer to keep them separate commands, but that's just my opinion. I'm not sure which way is best

**Comment by fabeschan** [comment]
> Usability: Before and after each of the commands runs, there should be more stuff printed to stdout saying what's going on and what to do next.

Added more instructions to stdout where appropriate. Please review and let me know if there are more things I can mention @percyliang

**Comment by fabeschan** [comment]
> update test-cli.py to test this functionality

Not exactly sure how to do this since some of the commands drop the user to a shell inside a container. Are there ways to run commands in the inner shell with python?

File: codalab/lib/bundle_cli.py (hunk 2 above)
Comment (fabeschan) [reply codalab/lib/bundle_cli.py]: @percyliang

**Pull request closed without merging**